map/bookmark_helpers.cpp

```diff
@@ -199,10 +199,57 @@ void ValidateKmlData(std::unique_ptr<kml::FileData> & data)
   if (!data)
     return;
 
+  // Filter points with the duplicated coordinates.
+  /// @todo(KK): This code is a temporary solution for the filtering the duplicated points in KMLs.
+  /// When the deserealizer reads the data from the KML that uses <gx:Track>
+  /// as a first step will be parsed the list of the timestamps <when> and than the list of the coordinates <gx:coord>.
+  /// So the filtering can be done only when all the data is parsed.
   for (auto & t : data->m_tracksData)
   {
     if (t.m_layers.empty())
       t.m_layers.emplace_back(kml::KmlParser::GetDefaultTrackLayer());
+
+    kml::MultiGeometry validGeometry;
+
+    auto const & geometry = t.m_geometry;
+    for (size_t lineIndex = 0; lineIndex < geometry.m_lines.size(); ++lineIndex)
+    {
+      auto const & line = geometry.m_lines[lineIndex];
+      auto const & timestamps = geometry.m_timestamps[lineIndex];
+
+      if (line.empty())
+      {
+        LOG(LWARNING, ("Empty line in track:", t.m_name[kml::kDefaultLang]));
+        continue;
+      }
+
+      bool const hasTimestamps = geometry.HasTimestampsFor(lineIndex);
+      if (hasTimestamps && timestamps.size() != line.size())
+        MYTHROW(kml::DeserializerKml::DeserializeException, ("Timestamps count", timestamps.size(), "doesn't match points count", line.size()));
+
+      validGeometry.m_lines.emplace_back();
+      validGeometry.m_timestamps.emplace_back();
+
+      auto & validLine = validGeometry.m_lines.back();
+      auto & validTimestamps = validGeometry.m_timestamps.back();
+
+      for (size_t pointIndex = 0; pointIndex < line.size(); ++pointIndex)
+      {
+        auto const & currPoint = line[pointIndex];
+
+        // We don't expect vertical surfaces, so do not compare heights here.
+        // Will get a lot of duplicating points otherwise after import some user KMLs.
+        // https://github.com/organicmaps/organicmaps/issues/3895
+        if (validLine.empty() || !AlmostEqualAbs(validLine.back().GetPoint(), currPoint.GetPoint(), kMwmPointAccuracy))
+        {
+          validLine.push_back(currPoint);
+          if (hasTimestamps)
+            validTimestamps.push_back(timestamps[pointIndex]);
+        }
+      }
+    }
```

Comment:
You're adding new code to existing function `ValidateKmlData`. For each element of `data.m_tracksData`, you eliminate points that are too close to each other.

I suggest creating a new function `RemovePointsTooCloseFromLine` for this operation, if possible. You might even add a parameter for the minimal distance (instead of `kMwmPointAccuracy`).

This would allow you to add unit testing for this function.

Reply:
Thanks! I'll add this change as part of the #9140